### ISSUE TITLE
Environment specific tool specs when using Dify via its API SDK

### ISSUE THREAD

**Issue by why-el** [langgenius/dify#11816]
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

I am not sure if this is provided already. We want to use Dify and trigger various workflows via its API, for instance, a chat embedded in our nodejs application. The workflow uses custom tools which hit our internal APIs, but the custom tools are not "environment-aware", so that developers can use them locally to hit their own localhost servers (for instance, via reverse proxies), while our production hits the production servers. 

It seems like the baked assumption for tools is that they are always supposed to hit production endpoints, which affects cycles of development greatly.

### 3. Can you help us with this feature?

- [ ] I am interested in contributing to this feature.

**Comment by crazywoola** [comment]
Do you want to have several instances eg staging, production and dev and they can access different environment separately?
I think you can do this in several ways.
Like postman and other tools we provide https://docs.dify.ai/guides/workflow/variables you can define a variable in inputs and use if/else to distinguish the environments, or you can simply deploy multiple instances among the teams.

**Comment by why-el** [comment]
Not exactly. This is about custom tools, for instance ones that hit our own internal APIs. Consider the following: 

- You wanna incorporate a custom tool that requests say user orders from your application. 
- You add its OpenAPI spec, and save the custom tool. 

Now you can use it, but the custom tool, during development, needs to hit `localhost`, but in production, it needs to hit our deployed application. How do you do this in a developer ergonomic way? I don't think Dify supports it. 

Furthermore, fiddling with OpenAPI specification is itself not super ergonomic. I wonder if we can create custom tools via the API?

**Comment by crazywoola** [comment]
Do you have any suggestions? I think there might be a way that you could add 3 tools, and each of them would account for the environment. Eg, dev-tool-spec, stg-tool-spec, prod-tool-spec.

**Comment by why-el** [comment]
You can do that yeah, but then how do you keep them in sync? do you update all? and do so manually? 

I think it would be ideal if Dify allows dynamic tool creation via an exposed API, similar to this for instance: https://humanloop.com/docs/v5/api-reference/tools/upsert.

**Comment by crazywoola** [comment]
Yeah, this seems doable, however we aim to offer orchestrated application instead of by exposing almost everything.

**Comment by perzeuss** [comment]
I thought about this a while ago. In theory, it's not just related to environments. Maybe you just want to use multiple accounts from a 3rd party provider, for example, but we can only configure a tool once with the credentials of one account. 
My suggestion would be to introduce a credentials store. If you drag a tool into a workflow, you can then use a select with autocomplete to find the appropriate access data from the credentials. This way, the credentials for a tool are not defined once for the workspace, but can be defined in every workflow. You can then also have these credentials selected dynamically using an input variable by reading the "credentials" field of the tools from, for example. a variable that is then called production_api_key. In contrast to environment variables that you have to configure separately in each workflow, you could define these credentials once across the workspace and then reference them in workflows with their ID. This also has the advantage that you only have to configure the credentials once. You can then build integrations for this later so that you can access/sync credentials from a vault in Dify, from providers like 1password.

What you would have to implement for this would be two things:
- UI, API and DB table for the credentials (is a standalone feature)
- Tool nodes must be able to accept the credential IDs instead of reading the credentials configured globally for the workspace.
